File: Concurrency/Scala/DiningPhilosophers.scala


import java.util.concurrent.locks.{ Lock, ReentrantLock }

object DiningPhilosophers {
	def main(args: Array[String]) { new DiningPhilosophers }
}

class DiningPhilosophers {

	val forks = for (i <- 0 to 5) yield new ReentrantLock(true)
	val philosophers = for (i <- 0 to 5) yield new Philosopher(i, forks(i), forks(if (i == 5) 0 else i + 1))
	philosophers.foreach((p: Philosopher) => new Thread(p).start)

	class Philosopher(id: Int, left: ReentrantLock, right: ReentrantLock) extends Runnable {
		def think = Thread.sleep(2000)
		def eating = think

		def eat {
			left.lock
			right.lock
			println(id + " Eating")
			eating
			left.unlock
			right.unlock
		}

		def eat2 {
			right.lock
			left.lock
			println(id + " Eating")
			eating
			right.unlock
			left.unlock
		}

		def run {
			while (true) {
				think
				if (id == 5) eat2 else eat
			}
		}

	}

}

File: Concurrency/C++/diningPhilosophers.cpp

#include <iostream>
#include <vector>
#include <thread>
#include <mutex>
#include <condition_variable>
#include <memory>
#include <chrono>

std::chrono::milliseconds oneMilliSecond(1);

class Philosopher {
public:

	Philosopher(int id, std::shared_ptr<std::mutex> left, std::shared_ptr<std::mutex> right) :
		id{id}, left{left}, right{right}
	{
	}

	void start()
	{
		while (true) {
			if (id != 5) {
				eat();
				think();
			} else {
				eat2();
				think();
			}
		}
	}

	void eat()
	{
		std::lock_guard<std::mutex> guard(*left.get());
		std::lock_guard<std::mutex> guard2(*right.get());
		std::cout << "Philosopher: " << id << " is eating food. \n";
		std::this_thread::sleep_for(100 * oneMilliSecond);
	}

	void eat2()
	{
		std::lock_guard<std::mutex> guard(*right.get());
		std::lock_guard<std::mutex> guard2(*left.get());
		std::cout << "Philosopher: " << id << " is eating food. \n";
		std::this_thread::sleep_for(100 * oneMilliSecond);
	}

	void think()
	{
		std::this_thread::sleep_for(10 * oneMilliSecond);
	}

	int id;
	std::shared_ptr<std::mutex> left;
	std::shared_ptr<std::mutex> right;

};


int main()
{

	std::vector<std::shared_ptr<std::mutex>> forks{
		std::shared_ptr<std::mutex>{new std::mutex},
		std::shared_ptr<std::mutex>{new std::mutex},
		std::shared_ptr<std::mutex>{new std::mutex},
		std::shared_ptr<std::mutex>{new std::mutex},
		std::shared_ptr<std::mutex>{new std::mutex}
	};

	std::vector<Philosopher> philosophers{
		Philosopher{1, forks[0], forks[1]},
		Philosopher{2, forks[1], forks[2]},
		Philosopher{3, forks[2], forks[3]},
		Philosopher{4, forks[3], forks[4]},
		Philosopher{5, forks[4], forks[0]}};

	std::vector<std::thread> threads;

	for (Philosopher p : philosophers)
		threads.push_back(std::thread(&Philosopher::start, p));

	for (std::thread &t : threads)
		t.join();

	return 0;
}
